### PR TITLE
feat: add configurable cache directory location

### PR DESCRIPTION
**Pull request by Louden7**
## What

Adds support for a custom cache directory location, allowing users to override the default cache path via cli option or toml configuration.

## Why

complexipy stores per-function complexity data between runs to power delta/diff comparisons. By default this cache lives in `.complexipy_cache/` inside your invocation directory. Instead of the default location you can redirect it to any directory which is useful for CI, chared cache directories or avoiding polluting the project tree.

## Changes

- Introduced functionality allowing users to modify the cache directory location. a4232786ed4cef03933138f4a85a789b4bcef481 
- Included documentation on configuring the custom cache directory location feature. 844809c72566023ed11ce9280ea7df4729ba32b9
- Wrote unit tests for feature coverage. 4372e7bdf2a7a2419314cad65d7f8d0a2fb5fbde & 21ac528bdae78572013db099d4e1e206249076e7

## Checklist

- [x] Tests pass (`uv run pytest tests/`)
- [x] Code is formatted (`uv run ruff check .` and `uv run ruff format .`)
- [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/): `type(scope): description`

### CODE REVIEW

File: docs/usage-guide.md [hunk 1]
```diff
@@ -383,6 +386,40 @@ section to the same configuration file:
...
+=== "pyproject.toml"
+
+    ```toml
+    cache-dir = ".cache/complexipy"
+    ```
+
+=== ".complexipy.toml"
+
+    ```toml
+    cache-dir = ".cache/complexipy
```
Comment (rohaquinlop) [inline docs/usage-guide.md]: Missing closing quote makes this invalid TOML, copying it breaks config parsing. `tomllib` rejects it with `Unterminated string`. Fix to `cache-dir = ".cache/complexipy"`.

Comment (rohaquinlop) [inline docs/usage-guide.md]: Missing closing quote makes this invalid TOML — copying it breaks config parsing. `tomllib` rejects it with `Unterminated string`. Fix to `cache-dir = ".cache/complexipy"`.

File: complexipy/utils/cache.py [hunk 2]
```diff
@@ -287,7 +294,7 @@ def _remove_legacy_cache_files(cache_dir: Path) -> None:
 
 def _ensure_cache_dir_and_supporting_files(cache_dir: Path) -> None:
     """Create the cache directory and support files used by cache tools."""
-    cache_dir.mkdir(exist_ok=True)
+    cache_dir.mkdir(parents=True, exist_ok=True)
```
Comment (rohaquinlop) [inline complexipy/utils/cache.py:305]: This `mkdir(parents=True)` change enables pointing `--cache-dir` at any pre-existing directory, but `_remove_legacy_cache_files` (line 287) unconditionally unlinks every `*.json` in the cache dir root on each successful run. That was safe while the dir was always the tool-owned `.complexipy_cache/`, but `--cache-dir .` or any directory holding JSON files would silently lose them.

Fix by scoping cleanup to the legacy filename pattern produced by `_build_cache_key` (32 lowercase hex chars + `.json`). Add the pattern with the other module constants in `cache.py` and filter the loop — `re` is already imported:

```python
LEGACY_CACHE_FILE_PATTERN = re.compile(r"^[0-9a-f]{32}\.json$")


def _remove_legacy_cache_files(cache_dir: Path) -> None:
    for legacy_cache_file in cache_dir.glob("*.json"):
        if not LEGACY_CACHE_FILE_PATTERN.match(legacy_cache_file.name):
            continue
        try:
            legacy_cache_file.unlink()
        except OSError:
            pass
```

File: complexipy/utils/cache.py [hunk 3]
```diff
@@ -32,10 +32,17 @@
 """
 
 
+def _resolve_cache_dir(invocation_path: str, cache_dir: Optional[str]) -> Path:
+    if cache_dir:
+        return Path(cache_dir)
```
Comment (rohaquinlop) [inline complexipy/utils/cache.py:38]: `cache_dir` comes straight from TOML without type validation, so `cache-dir = 123` in a config file reaches `Path(123)` here and crashes with an uncaught `TypeError` traceback. A wrong TOML type is a user config error, so fail with a clear message and exit code 1 — the same pattern `get_argument_value` already uses (`typer.echo` + `raise Exit`).

Apply this validation in `complexipy/utils/toml.py` by replacing the block added in this PR (lines 282–284):

```python
    cache_dir = cli_args.get("cache_dir")
    if cache_dir is None and toml_config is not None:
        cache_dir = toml_config.get("cache-dir")
    if cache_dir is not None and not isinstance(cache_dir, str):
        typer.echo("The 'cache-dir' option must be a string in the TOML config file")
        raise Exit(code=1)
    result["cache_dir"] = cache_dir
```

`typer` and `Exit` are already imported in `toml.py`. Do not silently coerce: a number for a path is a config bug and must be reported.

File: docs/usage-guide.md [hunk 4]
```diff
@@ -383,6 +386,40 @@ section to the same configuration file:
   the enforcement still applies. Fetch the branch or pass `--diff <ref>`
   with an existing reference instead.
 
+## Cache Configuration
+
+complexipy stores per-function complexity data between runs to power
+delta/diff comparisons. By default this cache lives in
+`.complexipy_cache/` inside your invocation directory. Instead of the
+default location you can redirect it to any directory which is useful
+for CI, chared cache directories or avoiding polluting the project
```
Comment (rohaquinlop) [inline docs/usage-guide.md]: "chared cache directories" → "shared cache directories".

File: tests/test_config.py [hunk 5]
```diff
@@ -37,6 +37,7 @@ def test_default_complexity(self):
             check_script=None,
             no_ignore=None,
             report_ignored=None,
+            cache_dir=None,
```
Comment (rohaquinlop) [inline tests/test_config.py:40]: All existing tests just pass `cache_dir=None`. The new `get_arguments_value` branch (TOML `cache-dir` resolution and CLI-over-TOML precedence) is never exercised — a regression there would go unnoticed. Add these two tests following the existing direct-call pattern:

```python
    def test_toml_cache_dir(self):
        cfg = resolve_config(
            {"cache-dir": "/tmp/cache"},
            paths=["."],
            max_complexity_allowed=None,
            snapshot_create=None,
            snapshot_ignore=None,
            quiet=None,
            ignore_complexity=None,
            failed=None,
            color=None,
            sort=None,
            output_format=None,
            output=None,
            diff=None,
            diff_only=None,
            staged=None,
            top=None,
            plain=None,
            suggest_refactors=None,
            exclude=None,
            check_script=None,
            no_ignore=None,
            report_ignored=None,
            cache_dir=None,
        )
        assert cfg.cache_dir == "/tmp/cache"

    def test_cli_cache_dir_overrides_toml(self):
        cfg = resolve_config(
            {"cache-dir": "/tmp/cache"},
            paths=["."],
            max_complexity_allowed=None,
            snapshot_create=None,
            snapshot_ignore=None,
            quiet=None,
            ignore_complexity=None,
            failed=None,
            color=None,
            sort=None,
            output_format=None,
            output=None,
            diff=None,
            diff_only=None,
            staged=None,
            top=None,
            plain=None,
            suggest_refactors=None,
            exclude=None,
            check_script=None,
            no_ignore=None,
            report_ignored=None,
            cache_dir="/tmp/cli",
        )
        assert cfg.cache_dir == "/tmp/cli"
```

File: tests/test_cache.py [hunk 6]
```diff
@@ -215,3 +215,100 @@ def test_cache_failure_does_not_break_functionality(self, tmp_path: Path):
...
+        )
+
+        files, _ = _complexipy.main([str(test_file)], False, [])
+
+        custom_cache_dir = tmp_path / ".cache" / "complexipy"
+        remember_previous_functions(
+            invocation_path=str(tmp_path),
+            targets=[str(test_file)],
+            files_complexities=files,
+            cache_dir=custom_cache_dir,
```
Comment (rohaquinlop) [inline tests/test_cache.py]: `cache_dir=custom_cache_dir` passes a `Path`, but the parameter is annotated `Optional[str]` and the sibling test uses `str(custom_cache_dir)`. Change this line to `cache_dir=str(custom_cache_dir)`. Also replace the fixture string in the three new tests (`"def example():\n   if True\n       return 1\n  return 0\n"` at lines 223, 254, 287) with the valid fixture already used in this file:

```python
"def example():\n    if True:\n        return 1\n    return 0\n"
```

**Review by rohaquinlop — CHANGES_REQUESTED** [review]
Feature verified end-to-end: CLI flag, TOML config, nested dirs, CLI-over-TOML precedence, and the default path all behave as documented. CI is green. Six inline comments below request changes.

One note that cannot be anchored inline: the repo keeps EN and ES docs in sync. Update `docs/es/usage-guide.md` too — add `cache-dir = ".complexipy_cache"` to the three config example blocks and add a Spanish "Configuración de Cache" section before `## API de Python`, mirroring the new EN section.

**Comment by rohaquinlop** [comment]
@Louden7 thank you for your contribution, I've used AI to automate the request changes comments but I've manually reviewed this PR

File: docs/usage-guide.md (hunk 1 above)
Comment (Louden7) [reply docs/usage-guide.md]: 167c693456dc35273d8ec2c43677c0828fd416e2

File: docs/usage-guide.md (hunk 4 above)
Comment (Louden7) [reply docs/usage-guide.md]: 167c693456dc35273d8ec2c43677c0828fd416e2

File: complexipy/utils/cache.py (hunk 2 above)
Comment (Louden7) [reply complexipy/utils/cache.py:305]: e83606b9cd560ad7ee6040a81add2f52182c0908

File: complexipy/utils/cache.py (hunk 3 above)
Comment (Louden7) [reply complexipy/utils/cache.py:38]: ad28eddbc3ee0917c424b60ebe26c44b953ae09c

File: tests/test_cache.py (hunk 6 above)
Comment (Louden7) [reply tests/test_cache.py]: cf81c01e475de902a7fc1a5cd4b71253c884fe95

File: tests/test_config.py (hunk 5 above)
Comment (Louden7) [reply tests/test_config.py:40]: a576dd440b181737e04eabe9993b42f4ed853a1c

**Comment by Louden7** [comment]
Implemented the changes above. Resolved _outdated_ and added a comment with a link to the respective commit. Let me know if there's anything else that needs fixing!

File: complexipy/utils/cache.py [hunk 7]
```diff
@@ -65,7 +73,7 @@ def remember_previous_functions(
     }
     _prune_cache_store(cache_store)
     if _persist_cache(cache_file, cache_store):
-        _remove_legacy_cache_files(cache_dir)
+        _remove_legacy_cache_files(cache_dir_path)
```
Comment (rohaquinlop) [inline complexipy/utils/cache.py]: Round-1 gating fixed the crash-scale version of this, but the cleanup still runs wherever `--cache-dir` points. Legacy `<32-hex>.json` files were only ever written under the default `.complexipy_cache/`, yet with a custom dir the regex now deletes any foreign tool's JSON that happens to match — and 32-hex `.json` is a common content-addressed naming scheme (MD5-keyed caches), so pointing `--cache-dir` at a shared/pre-existing directory can still silently delete data it does not own. `_migrate_legacy_cache_entry` (line 66) has the same provenance problem (read-only, but pointless outside the default location since legacy files never existed there). Gate both on the default location:

```python
    cache_dir_path = _resolve_cache_dir(invocation_path, cache_dir)
    is_default_cache_location = (
        cache_dir_path == Path(invocation_path) / CACHE_DIR_NAME
    )
```

then wrap both call sites:

```python
    if is_default_cache_location:
        _migrate_legacy_cache_entry(cache_dir_path, cache_store, cache_key)
```

```python
    if _persist_cache(cache_file, cache_store):
        if is_default_cache_location:
            _remove_legacy_cache_files(cache_dir_path)
```

File: tests/test_config.py [hunk 8]
```diff
@@ -573,8 +592,65 @@ def test_plain_and_quiet_raises(self):
...
+            suggest_refactors=None,
+            exclude=None,
+            check_script=None,
+            no_ignore=None,
+            report_ignored=None,
+            cache_dir=None,
+        )
+        assert cfg.cache_dir == "/tmp/cache"
+
+    def test_cli_cache_dir_overrides_toml(self):
```
Comment (rohaquinlop) [inline tests/test_config.py:626]: The validation added in `complexipy/utils/toml.py` (lines 286–289, `Exit(code=1)` for non-string values) is the only new error-handling code in this PR and no test exercises it — a regression to a raw `TypeError` would go unnoticed. Add one test after the precedence tests, following the existing `pytest.raises(Exception)` style:

```python
    def test_non_string_toml_cache_dir_raises(self):
        with pytest.raises(Exception):
            resolve_config(
                {"cache-dir": 123},
                paths=["."],
                max_complexity_allowed=None,
                snapshot_create=None,
                snapshot_ignore=None,
                quiet=None,
                ignore_complexity=None,
                failed=None,
                color=None,
                sort=None,
                output_format=None,
                output=None,
                diff=None,
                diff_only=None,
                staged=None,
                top=None,
                plain=None,
                suggest_refactors=None,
                exclude=None,
                check_script=None,
                no_ignore=None,
                report_ignored=None,
                cache_dir=None,
            )
```

File: complexipy/utils/toml.py [hunk 9]
```diff
@@ -279,4 +279,14 @@ def get_arguments_value(
         toml_config, "exclude", cli_args.get("exclude"), []
     )
 
+    cache_dir = cli_args.get("cache_dir")
+    if cache_dir is None and toml_config is not None:
+        cache_dir = toml_config.get("cache-dir")
```
Comment (rohaquinlop) [inline complexipy/utils/toml.py:284]: `cache-dir = ""` passes the `isinstance(str)` check below, then `_resolve_cache_dir` treats `""` as falsy and silently uses the default location. That implicit contract is undocumented and untested; a refactor flipping the truthiness check would change behavior silently. Prefer rejecting empty strings here with the same message pattern as the type error:

```python
    if cache_dir is not None and not isinstance(cache_dir, str):
        ...
    if cache_dir is not None and not cache_dir.strip():
        typer.echo(
            "The 'cache-dir' option cannot be an empty string in the TOML config file"
        )
        raise Exit(code=1)
```

or, if you prefer keeping the fallback, lock it with an end-to-end test asserting `cfg.cache_dir == ""` resolves to the default cache directory.

File: tests/test_cache.py [hunk 10]
```diff
@@ -215,3 +215,100 @@ def test_cache_failure_does_not_break_functionality(self, tmp_path: Path):
...
+
+        files, _ = _complexipy.main([str(test_file)], False, [])
+
+        custom_cache_dir = tmp_path / "custom_cache"
+        remember_previous_functions(
+            invocation_path=str(tmp_path),
+            targets=[str(test_file)],
+            files_complexities=files,
+            cache_dir=str(custom_cache_dir),
+        )
```
Comment (rohaquinlop) [inline tests/test_cache.py:280]: `test_custom_cache_dir` asserts creation and population but never calls `remember_previous_functions` a second time against the custom location, so a bug that writes to the custom dir but reads from the default would pass. Add the round-trip assertion before the final checks:

```python
        files_again, _ = _complexipy.main([str(test_file)], False, [])
        previous = remember_previous_functions(
            invocation_path=str(tmp_path),
            targets=[str(test_file)],
            files_complexities=files_again,
            cache_dir=str(custom_cache_dir),
        )
        assert previous is not None, (
            "Second run should read the previous map from the custom dir"
        )
        assert len(previous) == 1
```

**Review by rohaquinlop — CHANGES_REQUESTED** [review]
Re-review of head `a576dd4`.

**Round-1 status:** all six inline comments verified fixed — TOML quote and 'shared' typo (`167c693`), legacy cleanup gated to `LEGACY_CACHE_FILE_PATTERN` (`e83606b`), TOML type validation with `Exit(1)` (`ad28edd`), precedence tests (`a576dd4`), fixture + `Path→str` cast (`cf81c01`). Local run green (312 tests, ruff, ruff format, ty) and CI green. Nice work.

**Still blocking (from round 1):** `docs/es/usage-guide.md` is not updated. The repo keeps EN and ES docs in sync — add `cache-dir = ".complexipy_cache"` to the three config example blocks and insert a Spanish "Configuración de Cache" section before `## API de Python`, mirroring the new EN section.

Four new inline comments this round: one safety hardening (scope legacy cleanup/migration to the default cache location), three test-gap/edge-case items. No blockers beyond that.

File: complexipy/utils/cache.py (hunk 7 above)
Comment (Louden7) [reply complexipy/utils/cache.py]: 317e56a8eb1c52daee56e9ce486b26aab4c3ae54

File: tests/test_config.py (hunk 8 above)
Comment (Louden7) [reply tests/test_config.py:626]: 58e7383455baf204595fe4054bd90c18cbd6695e

File: complexipy/utils/toml.py (hunk 9 above)
Comment (Louden7) [reply complexipy/utils/toml.py:284]: 9c7198268b87b8adcc62e722993a6acb7fac6430

File: tests/test_cache.py (hunk 10 above)
Comment (Louden7) [reply tests/test_cache.py:280]: 2d0ea3fdeb9297d9e7f592eead28873deb4074dd

**Comment by Louden7** [comment]
Implemented changes requested. Also did my best with AI and Google Translate for the en to es docs addition [ceb3a38b57911d7d878ef765b01fc338a31c30f7]

**Review by rohaquinlop — APPROVED** [review]
